### PR TITLE
Bugfixes

### PR DESCRIPTION
**Pull request by AJenbo**

### CODE REVIEW

**Comment by Predelnik** [comment]
Does any of those fixes ends up producing different dungeon for certain seeds?
And have we actually checked how monster interaction with fire wall changes?

**Comment by AJenbo** [comment]
From what I have seen the broken seeds are unplayable to some extend. But we definitely should play test this before it goes in, especially before reaching 1.0.0

**Comment by AJenbo** [comment]
@glebm going to rebase this and merge

File: Source/drlg_l2.cpp [hunk 1]
```diff
@@ -903,8 +903,9 @@ void DRLG_L2PlaceRndSet(BYTE *miniset, int rndper)
 			if (found == TRUE) {
 				for (yy = sy - sh; yy < sy + 2 * sh && found == TRUE; yy++) {
 					for (xx = sx - sw; xx < sx + 2 * sw; xx++) {
-						// BUGFIX: yy and xx can go out of bounds
-						if (dungeon[xx][yy] == miniset[kk]) {
+						// BUGFIX: yy and xx can go out of bounds (fixed)
+						if(0 <= xx && xx < DMAXX && 0 <= yy && yy < DMAXY &&
```
Comment (glebm) [inline Source/drlg_l2.cpp]: Seems like a better place to fix this is adding `std::min/max` to the loop initializers and conditions

File: Source/drlg_l3.cpp [hunk 2]
```diff
@@ -1065,8 +1065,8 @@ void DRLG_L3Pool()
 			for (j = duny - totarea; j < duny + totarea; j++) {
 				for (i = dunx - totarea; i < dunx + totarea; i++) {
 					// BUGFIX: In the following swap the order to first do the
-					// index checks and only then access dungeon[i][j]
-					if (dungeon[i][j] & 0x80 && j >= 0 && j < DMAXY && i >= 0 && i < DMAXX) {
+					// index checks and only then access dungeon[i][j] (fixed)
+					if (j >= 0 && j < DMAXY && i >= 0 && i < DMAXX && dungeon[i][j] & 0x80) {
```
Comment (glebm) [inline Source/drlg_l3.cpp]: Same, `std::min/max` in the `for` statements seems like a cleaner fix

File: Source/drlg_l4.cpp [hunk 3]
```diff
@@ -1445,8 +1445,9 @@ void uShape()
 				hallok[j] = FALSE;
 			}
 			if (dung[i][j] == 1) {
-				// BUGFIX: check that i + 1 < 20 and j + 1 < 20
-				if (dung[i][j + 1] == 1 && dung[i + 1][j + 1] == 0) {
+				// BUGFIX: check that i + 1 < 20 and j + 1 < 20 (fixed)
+				if (i + 1 < 20 && j + 1 < 20 &&
```
Comment (glebm) [inline Source/drlg_l4.cpp]: nit for the future: `20` should probably be a constant like in the other `drlg` files

Comment (AJenbo) [reply Source/drlg_l4.cpp]: I guess it could be `DMAXX / 2`? Any way that should be done upstream :)

**Comment by AJenbo** [comment]
@glebm would you be willing to double check that I did the min/max correct :)

File: Source/drlg_l3.cpp [hunk 4]
```diff
@@ -792,11 +793,11 @@ static void DRLG_L3Pool()
 				found = TRUE;
 			}
 			poolchance = random_(0, 100);
-			for (j = duny - totarea; j < duny + totarea; j++) {
-				for (i = dunx - totarea; i < dunx + totarea; i++) {
+			for (j = std::max(0, std::min(DMAXY - 1, duny - totarea)); j < duny + totarea; j++) {
```
Comment (glebm) [inline Source/drlg_l3.cpp]: This should be more like:

```c++
for (j = std::max(duny - totarea, 0); j < std::min(duny + totarea, DMAXY); ++j) {
```

Similar in other places

Comment (AJenbo) [reply Source/drlg_l3.cpp]: Ok how about now :)

**Comment by AJenbo** [comment]
AI appears unaffected so I'm ok with merging, but MSVC is uhappy about the header so not sure we can use min/max, but since one value is static we can use `<?:` instead.

**Comment by glebm** [comment]
@AJenbo Looks like it's because we include the `windows.h` header which defines min/max macros. According to this answer, we should `#define NOMINMAX` on MSVC:
https://stackoverflow.com/questions/5004858/why-is-stdmin-failing-when-windows-h-is-included

**Comment by AJenbo** [comment]
For Source we don't include windows.h as it will conflict with miniwin (SourceS), i think whats happening is that it's pulling in some other headers that conflict with miniwin. If your ok with using `<?:` instead then I would like to just merge that rather then fight header hell.

**Comment by glebm** [comment]
Sounds good, I can investigate the header hell later. I think one of the dependencies might be including `windows.h`.

**Comment by AJenbo** [comment]
Looks like Travis is overloaded again, guess we will have to fly blind, but at least we only use it for Mac builds.

**Pull request merged**